src/gd_avif.c

```diff
@@ -0,0 +1,667 @@
+    goto cleanup;
+  }
+
+  // Encode the image in AVIF format
+
+  avifEncoder * encoder = avifEncoderCreate();
+  int quantizerQuality = quality == DEFAULT_QUALITY ? 
+                         DEFAULT_QUANTIZER : quality2Quantizer(quality);
+
+  encoder->minQuantizer = quantizerQuality;
+  encoder->maxQuantizer = quantizerQuality;
+  encoder->minQuantizerAlpha = quantizerQuality;
+  encoder->maxQuantizerAlpha = quantizerQuality;
+  encoder->speed = speed;
+  setEncoderTilesAndThreads(encoder, &rgb);
+
+  result = avifEncoderAddImage(encoder, avifIm, 1, AVIF_ADD_IMAGE_FLAG_SINGLE); //TODO: why 1?
+  if ((failed = isAvifError(result, "Could not encode image"))) {
+    goto cleanup;
+  }
+
+  result = avifEncoderFinish(encoder, &avifOutput);
+  if ((failed = isAvifError(result, "Could not finish encoding"))) {
+    goto cleanup;
+  }
+
+  // Write the AVIF image bytes to the GD ctx.
+
+  gdPutBuf(avifOutput.data, avifOutput.size, outfile);
+
+  cleanup:
+    if (rgb.pixels) {
+      avifRGBImageFreePixels(&rgb);
+    }
+
+    if (encoder) {
+      avifEncoderDestroy(encoder);
+    }
+
+    if (avifOutput.data) {
+      avifRWDataFree(&avifOutput);
+    }
+
+    return failed;
+}
+
+/*****************************************************************************
+ * 
+ *                            HELPER FUNCTIONS
+ * 
+ *****************************************************************************/
+
+/* Convert the quality param we expose to the quantity params used by libavif.
+   The *Quantizer* params values can range from 0 to 63, with 0 = highest quality and 63 = worst.
+   We make the scale 0-100, and we reverse this, so that 0 = worst quality and 100 = highest.
+
+   Values below 0 are set to 0, and values below MAX_QUALITY are set to MAX_QUALITY.
+*/
+
+static int quality2Quantizer(int quality) {
+  int clampedQuality = AVIF_CLAMP(quality, 0, MAX_QUALITY);
+
+  float scaleFactor = (float) AVIF_QUANTIZER_WORST_QUALITY / (float) MAX_QUALITY;
+
+  return round(scaleFactor * (MAX_QUALITY - clampedQuality));
+}
+
+/*
+  From gd_png.c:
+    convert the 7-bit alpha channel to an 8-bit alpha channel.
+    We do a little bit-flipping magic, repeating the MSB
+    as the LSB, to ensure that 0 maps to 0 and
+    127 maps to 255. We also have to invert to match
+    PNG's convention in which 255 is opaque. 
+*/
+
+static uint8_t convertTo8BitAlpha(uint8_t originalAlpha) {
+  return originalAlpha == 127 ? 
+        0 : 
+        255 - ((originalAlpha << 1) + (originalAlpha >> 6));
+}
+
+/*
+   As of February 2021, this algorithm reflects the latest research on how many tiles
+   and threads to include for a given image size.
+   This is subject to change as research continues.
+ */
+
+static void setEncoderTilesAndThreads(avifEncoder * encoder, avifRGBImage * rgb) {
+  int imageArea, tiles, tilesLog2;
+  int tileRowsLog2, tileColumnsLog2, maxThreads;
+
+  imageArea = rgb->width * rgb->height;
```

Comment:
use `overflow2` before doing any multiplication

Reply:
Thanks. I saw that elsewhere in the code, but didn't implement it. Now I did. Integer overflow can be dangerous:

https://www.nytimes.com/2015/05/01/business/faa-orders-fix-for-possible-power-loss-in-boeing-787.html